content/posts/cross-compile-query-json.mdx

```diff
@@ -0,0 +1,252 @@
+where after the blur happens on the input a request needs to hit a backend and respond with the result. It's a massive improvement.
+
+### Benefits
+This allows any OCaml backend being able to run in a browser without much hassle, sharing code between backend and frontend has been a dream for a lot of Engineering teams.
+
+But not only share logic matters, here's a list of other upsides:
+
+- **Portability**, moving code from server to client or viceversa, sharing marshal/unmarshal code, easier contract testing.
+- **Familiarity**: Writting the same patterns would benefit new commers that need to learn less platform specific rules.
+- **Usage of OCaml's ecosystem**: Access to many libraries and ppxs and latests ocaml's features.
+
+### Challenges
+This solution has downsides as well, js_of_ocaml isn't the tool that solves all of your problems, actually there's no such tool.
+
+#### Bundlesize
+It's quite big compared to regular JavaScript applications.
+
+The entire playground is about 660kb. Includes the Monaco Editor (~356kb) and the rest is query-json and jsoo runtime (300kb).
+
+jsoo wasn't created with the same mentality as most Frontend developers solve their problems today and was borned 10 years ago to run some OCaml into a browser.
+
+#### Documentation
+One of my biggest complains about the OCaml community is the lack of quality documentation.
+
+Comming from the JavaScript community (which have more than 9 million devs) there's plenty of tutorials, examples, manuals, many StackOverflow resolved questions and this culture of the copy-paste driven development.
+
+That gives all sort of problems, but lowers the barrier to the language and the usage of many dependencies for new commers or not so passionate FP developers.
+
+It's a non-sense compare OCaml and JavaScript, but one of the biggest missing pieces are good documentation at all levels.
+
+#### Bridge between BuckleScript/Reason and jsoo
+It lacks some of the basics to enable compatibility with JavaScript codebases, modern build systems such as webpack and many bindings to other libraries.
+
+This isn't particular to jsoo, is more related to use jsoo and BuckleScript together. In order to run the jsoo artifact into the BuckleScript/Reason codebase, I needed to write bindings for it.
+
+```reason
+[@bs.module "../../_build/default/js/Js.bc.js"]
+external queryJson: (string, string) => result(string, string) = "run";
+```
+
+Using the [Result from OCaml](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Result.html) in the query-json's JavaScript entry-point and the [Result from BuckleScript](https://rescript-lang.org/docs/manual/latest/api/belt/result) in the Reason code I needed to write an unsafe bridge which transforms OCaml result to the internal representation of BuckleScript variant.
+
+There's probably better ways of achiving it since I made the implementation very unsafe [davesnx/query-json/js/Js.re](https://github.com/davesnx/query-json/blob/master/js/Js.re).
```

Comment:
```suggestion
There are probably better ways of achieving it since I made the implementation very unsafe [davesnx/query-json/js/Js.re](https://github.com/davesnx/query-json/blob/master/js/Js.re).
```